mlir/lib/Catalyst/Transforms/InlineNestedModules.cpp

```diff
@@ -0,0 +1,493 @@
+
+    SmallVector<Operation *> *_symbolTables;
+};
+
+static constexpr llvm::StringRef hasBeenRenamedAttrName = "catalyst.unique_names";
+
+LogicalResult RenameFunctionsPattern::match(Operation *op) const
+{
+    bool isSymbolTable = op->hasTrait<OpTrait::SymbolTable>();
+    bool hasBeenRenamed = op->hasAttr(hasBeenRenamedAttrName);
+    // TODO: isQnode
+    return isSymbolTable && !hasBeenRenamed ? success() : failure();
+}
+
+void RenameFunctionsPattern::rewrite(Operation *child, PatternRewriter &rewriter) const
+{
+    assert(child->hasTrait<OpTrait::SymbolTable>() && "child must be symbol table");
+    auto parent = child->getParentOp();
+    assert(parent && "parent must exist");
+    assert(parent->hasTrait<OpTrait::SymbolTable>() && "parent must be a symbol table");
+
+    SymbolTable childSymTab(child);
+    SmallVector<std::shared_ptr<SymbolTable>> tables;
+    SmallVector<SymbolTable *> raw_tables;
+    for (auto operation : *_symbolTables) {
+        std::shared_ptr<SymbolTable> sym(new SymbolTable(operation));
+        tables.push_back(sym);
+        raw_tables.push_back(sym.get());
+    }
+
+    // Yes, this is a triple nested loop, but it is faster
+    // than doing a post-order walk because we are not interested in
+    // nested regions nor blocks. Also, since child is expected to be a module
+    // then the number of regions is guaranteed to be one and the number of blocks
+    // is also guaranteed to be one.
+    for (auto &region : child->getRegions()) {
+        for (auto &block : region.getBlocks()) {
+            for (auto &op : block) {
+                if (!isa<SymbolOpInterface>(op))
+                    continue;
+
+                if (failed(childSymTab.renameToUnique(&op, raw_tables))) {
+                    // TODO: Check for error in one of the tests.
+                    op.emitError() << "Cannot rename operation";
+                    llvm_unreachable("Failed state");
+                }
+            }
+        }
+    }
+
+    rewriter.modifyOpInPlace(
+        child, [&] { child->setAttr(hasBeenRenamedAttrName, rewriter.getUnitAttr()); });
+}
+
+struct InlineNestedModule : public RewritePattern {
+    /// This overload constructs a pattern that matches any operation type.
+    InlineNestedModule(MLIRContext *context) : RewritePattern(MatchAnyOpTypeTag(), 1, context) {}
+
+    LogicalResult match(Operation *op) const override;
+    void rewrite(Operation *op, PatternRewriter &rewriter) const override;
+};
+
+LogicalResult InlineNestedModule::match(Operation *op) const
+{
+    bool isSymbolTable = op->hasTrait<OpTrait::SymbolTable>();
+    // TODO: isQnode
+    return isSymbolTable ? success() : failure();
+}
+
+void InlineNestedModule::rewrite(Operation *op, PatternRewriter &rewriter) const
+{
+    auto parent = op->getParentOp();
+    // Can't generalize getting a region other than the zero-th one.
+    rewriter.inlineRegionBefore(op->getRegion(0), &parent->getRegion(0).front());
```

Comment:
I think this line is why the lit tests' function order is reversed, since the functions that are visited and inlined later gets put in the front, e.g.

visiting order: f, g, h
module: (stuff)->(f, stuff)->(g, f, stuff)->(h, g, f, stuff)

Reply:
```suggestion
    rewriter.inlineRegionBefore(op->getRegion(0), &parent->getRegion(0).back());
```

Reply:
visiting order: f, g, h
module: (stuff)->(stuff, f)->(stuff, f, g)->(stuff, f, g, h)

Reply:
`back()` probably won't work if there's just one block in the region right? Or does it actually return the "end of container" iterator, in which case it would work?

Anyway I leave it to you to do this little surgery : )

Reply:
Oh no wait, `back()` would work:

visiting order: f, g, h
module: (stuff) [last block in module is "stuff", all inlines go right before it] -> (f, stuff) -> (f, g, stuff) -> (f, g, h, stuff)

Reply:
No, this is not it. The striked through comment is still correct. I think the only correct way is to not inline (but not erase) and then find where the module was, move all inlined functions there and then erase the module. Something like that.

Reply:
I think it's because the line below also needs to be adjusted, so 
```suggestion
    rewriter.inlineRegionBefore(op->getRegion(0), &parent->getRegion(0).back());
    rewriter.mergeBlocks(&parent->getRegion(0).back(), &parent->getRegion(0).front());
```

I just tested it and the order in frontend cancel_inverses test is no longer reversed.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1144/commits/92e61bc5863ea1a51e03745b2d231e71cfc3a63e I did this, which also preserves the order of other functions, not just the ones inlined. E.g., setup and teardown and potentially others.